tokio/src/sync/broadcast.rs

```diff
@@ -804,6 +817,50 @@ impl<T> Sender<T> {
         Arc::ptr_eq(&self.shared, &other.shared)
     }
 
+    /// A future which completes when the number of [Receiver]s subscribed to this `Sender` reaches
+    /// zero.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use futures::FutureExt;
+    /// use tokio::sync::broadcast;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx1) = broadcast::channel::<u32>(16);
+    ///     let mut rx2 = tx.subscribe();
+    ///
+    ///     tokio::spawn(async move {
+    ///         assert_eq!(rx1.recv().await.unwrap(), 10);
+    ///     });
+    ///
+    ///     let _ = tx.send(10);
+    ///     assert!(tx.closed().now_or_never().is_none());
+    ///
+    ///     let _ = tokio::spawn(async move {
+    ///         assert_eq!(rx2.recv().await.unwrap(), 10);
+    ///     }).await;
+    ///
+    ///     assert!(tx.closed().now_or_never().is_some());
+    /// }
+    /// ```
+    pub async fn closed(&self) {
+        self.shared.notify_rx_drop.notified().await;
+
+        poll_fn(|_| {
+            let tail = self.shared.tail.lock();
+
+            if tail.closed || tail.rx_cnt == 0 {
+                return Poll::Ready(());
+            }
+
+            drop(tail);
+            return Poll::Pending;
+        })
+        .await;
```

Comment:
If we now notify only once, when the last receiver drops, I think we can just `await` the Notified directly, right? I originally had this block because I notified on every `Rx` drop, so had to include an additional check to ensure that we had just dropped the last `Rx`. 

Relatedly, just for my own understanding - would this not work because the internal `Future` state machine is progressed to the point where we're already in the `poll_fn`, so future `closed.await`s wouldn't actually `await` the notified again?

Reply:
Yes, `Notified` doesn't have spurious wakeups in the scenario we are using it, so it should work. Well, except for the fact that we probably want it to return immediately if it's already closed when we call it. We can do it like this:
```rs
loop {
    let notified = self.shared.notify_rx_drop.notified();
    if self.is_closed() { return; }
    notified.await;
}
```

Reply:
Sure, done. For my own curiosity, what's the advantage of the loop above vs. a simple `self.shared.notify_rx_drop.notified().await`, if `Notified` doesn't suffer from spurious wakeups?

Note: I used `tail.closed` instead of `self.is_closed()` as I couldn't find an `is_closed` on `Sender`, which means an extra lock

Reply:
If you don't use a loop, then it changes the behavior of the test I requested in https://github.com/tokio-rs/tokio/pull/6685#discussion_r1770601675. Which behavior we want is up for discussion.